File: backend/app/schemas/sync.py
"""
Schemas Pydantic para sincronización offline
"""
from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel


class SyncOperation(BaseModel):
    """Schema para una operación de sincronización"""
    entity_type: str  # animal, finca, control_sanitario, etc.
    entity_id: int
    operation: str  # create, update, delete
    data: Optional[Dict[str, Any]] = None
    device_id: str
    local_timestamp: datetime
    sync_version: int


class SyncRequest(BaseModel):
    """Schema para solicitud de sincronización"""
    device_id: str
    last_sync: Optional[datetime] = None
    operations: List[SyncOperation] = []


class SyncConflict(BaseModel):
    """Schema para conflicto de sincronización"""
    entity_type: str
    entity_id: int
    server_version: int
    client_version: int
    server_data: Dict[str, Any]
    client_data: Dict[str, Any]
    conflict_resolution: str  # server_wins, client_wins, manual


class SyncResponse(BaseModel):
    """Schema para respuesta de sincronización"""
    success: bool
    synced_at: datetime
    conflicts: List[SyncConflict] = []
    updates_from_server: List[Dict[str, Any]] = []
    message: str


class SyncStats(BaseModel):
    """Estadísticas de sincronización"""
    last_sync: Optional[datetime] = None
    pending_operations: int
    synced_entities: int
    conflicts: int

File: frontend/src/api/sync.ts
import apiClient from './client';
import { getDeviceId, getLastSyncAt, setLastSyncAt } from '../offline/cache';

export interface SyncPullResponse {
  success: boolean;
  synced_at: string;
  updates_from_server: Array<{
    entity_type: string;
    entity_id: number;
    operation: string;
    data: Record<string, unknown>;
  }>;
  conflicts: unknown[];
  errors: string[];
  message: string;
}

export async function pullServerUpdates(): Promise<SyncPullResponse | null> {
  if (!navigator.onLine) return null;

  const lastSync = await getLastSyncAt();
  const response = await apiClient.post<SyncPullResponse>('/sync/sync', {
    device_id: getDeviceId(),
    last_sync: lastSync,
    operations: [],
  });

  if (response.data.synced_at) {
    await setLastSyncAt(response.data.synced_at);
  }

  return response.data;
}
